cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+        :param is_positive: defines the value to be used for your filtering column
+        """
+        df = cursor.execute(query).as_pandas()
+        df = (
+            df.sort_values(by=[self.config.primary_ref])
+            .reset_index()
+            .drop("index", axis=1)
+        )
+        df = (
+            # TODO: remove replace behavior after increasing data sample size
+            df.sample(n=sample_size, random_state=self.config.seed, replace=True)
+            .sort_values(by=[self.config.primary_ref])
+            .reset_index()
+            .drop("index", axis=1)
+        )
+
+        df[dependent_variable] = is_positive
+        return df
+
+    def _create_covariate_table(self, cursor: DatabaseCursor, schema: str):
+        """Creates a covariate table from the loaded toml config"""
+        # checks for primary & link ref being the same
+        source_refs = list(
+            {self.config.primary_ref, self.config.count_ref} - set([None])
+        )
+        pos_query = get_distinct_ids(source_refs, self.config.pos_source_table)
+        pos = self._get_sampled_ids(
+            cursor,
+            schema,
+            pos_query,
+            self.config.pos_sample_size,
+            self.config.dependent_variable,
+            1,
+        )
+        neg_query = get_distinct_ids(
+            source_refs,
+            self.config.neg_source_table,
+            join_id=self.config.primary_ref,
+            filter_table=self.config.pos_source_table,
+        )
+        neg = self._get_sampled_ids(
+            cursor,
+            schema,
+            neg_query,
+            self.config.neg_sample_size,
+            self.config.dependent_variable,
+            0,
+        )
+        cohort = pandas.concat([pos, neg])
+
+        # Replace table (if it exists)
+        # TODO - replace with timestamp prepended table
+        drop = get_drop_view_table(
+            f"{self.config.pos_source_table}_sampled_ids", "TABLE"
+        )
+        cursor.execute(drop)
+        ctas_query = get_ctas_query_from_df(
+            schema,
+            f"{self.config.pos_source_table}_sampled_ids",
+            cohort,
+        )
+        self.queries.append(ctas_query)
+        # TODO - replace with timestamp prepended table
+        drop = get_drop_view_table(self.config.target_table, "TABLE")
+        cursor.execute(drop)
+        dataset_query = get_create_covariate_table(
+            target_table=self.config.target_table,
+            pos_source_table=self.config.pos_source_table,
+            neg_source_table=self.config.neg_source_table,
+            primary_ref=self.config.primary_ref,
+            dependent_variable=self.config.dependent_variable,
+            join_cols_by_table=self.config.join_cols_by_table,
+            count_ref=self.config.count_ref,
+            count_table=self.config.count_table,
+        )
+        self.queries.append(dataset_query)
+
+    def generate_psm_analysis(self, cursor: object, schema: str):
```

Comment:
nit: `cursor: DatabaseCursor` unless you used `object` because you knowingly call a method not in the protocol and you didn't want to see a squiggly cursor in your IDE. In which case, let's sort that piece out (from discussion above).

Reply:
good catch - i believe this is a holdover from when i started and that PR hadn't been merged yet.